src/server/transaction.cc

```diff
@@ -94,12 +94,36 @@ bool CheckLocks(const DbSlice& db_slice, IntentLock::Mode mode, const KeyLockArg
 struct ScheduleContext {
   Transaction* trans;
   bool optimistic_execution = false;
+
+  std::atomic<ScheduleContext*> next{nullptr};
+
   std::atomic_uint32_t fail_cnt{0};
 
   ScheduleContext(Transaction* t, bool optimistic) : trans(t), optimistic_execution(optimistic) {
   }
 };
 
+struct ScheduleQ {
+  base::MPSCIntrusiveQueue<ScheduleContext> queue;
+
+  static constexpr size_t kSz = sizeof(queue);
+
+  char pad1[64];
+  atomic_bool armed{false};
+  char pad2[60];
+};
+
+void MPSC_intrusive_store_next(ScheduleContext* dest, ScheduleContext* next_node) {
+  dest->next.store(next_node, std::memory_order_relaxed);
+}
+
+ScheduleContext* MPSC_intrusive_load_next(const ScheduleContext& src) {
+  return src.next.load(std::memory_order_acquire);
+}
+
+// of shard_num arity.
+ScheduleQ* schedule_queues = nullptr;
```

Comment:
unique_ptr

Reply:
does not provide any value since we must delete manually anyways.